dotcom-rendering/src/web/components/DynamicFast.tsx

```diff
@@ -1,58 +1,427 @@
+ *       ^  -  ^  - standard - standard
+ * '2nd boosted' Container layout is
+ *      veryBig1  - veryBig2 (boosted)
+ *      big - big - standard - standard
+ *       ^  -  ^  - standard - standard
+ *       ^  -  ^  - standard - standard
+ *
+ * '3rd boosted' Container layout is
+ * (3rd can be boosted in combination with default, 1st boosted or 2nd boosted)
+ *       veryBig1     -   	veryBig2
+ *      big (boosted) - standard - standard
+ *            ^       - standard - standard
+ *            ^       - standard - standard
+ *            ^       - standard - standard
+ *
+ * '3rd boosted + 1 big' Container layout is
+ * (can be used in combination with default, 1st boosted or 2nd boosted)
+ *       veryBig1     -   	veryBig2
+ *      big (boosted) -    big   - standard
+ *            ^       -     ^    - standard
+ *            ^       -     ^    - standard
+ *            ^       - standard - standard
+ *            ^       - standard - standard
+ *
+ * There are also variations where a 'huge' is provided, or only 1 very big, where the top
+ * rows become just 1 'huge' card, instead of the 2 primaries shown in the layouts above.
+ */
+export const DynamicFast = ({
+	groupedTrails,
+	containerPalette,
+	showAge,
+}: Props) => {
+	// 'primaries' should hold up to 2 cards, selecting that which are given
+	// the 'largest' group.
+	let primaries: TrailType[] = [];
+	if (groupedTrails.huge.length > 0) {
+		// Where a huge is set, we only take 1 primary, leaving any
+		// 'veryBigs' to be displayed as 'bigs'
+		primaries = groupedTrails.huge.splice(0, 1);
+	} else if (groupedTrails.veryBig.length === 1) {
+		// This would display the same as 1 'huge'.
+		primaries = groupedTrails.veryBig.splice(0, 1);
+	} else if (groupedTrails.veryBig.length > 1) {
+		// We take a maximum of 2, remaining very bigs will be treated as 'big'
+		primaries = groupedTrails.veryBig.splice(0, 2);
+	}
+
+	// Put together all the bigs - only those not chosen as 'primaries' will be left over
+	const bigs = groupedTrails.huge
+		.concat(groupedTrails.veryBig, groupedTrails.big)
+		.slice(0, 4);
+
+	const firstBigBoosted = !!bigs[0]?.isBoosted;
+
+	// Gets display information
+	const standardsConfig = getStandardsDisplayConfig(
+		firstBigBoosted,
+		bigs.length,
+	);
+
+	let standards: TrailType[] = [];
+	if (firstBigBoosted && bigs.length > 1) {
+		// If the first big is boosted & there is an additional big card, we change
+		// we change to a 2 column layout. Any bigs but the first 2 should fall-through
+		// to being standards.
+		standards = [
+			// Take all but the first 2 bigs
+			...bigs.splice(2, bigs.length - 1),
+			...groupedTrails.standard,
+		];
+	} else if (standardsConfig.columns > 0) {
+		// We don't need to select any 'left over' bigs here as if there are
+		// 4 bigs there is no room for standards.
+		standards = groupedTrails.standard.splice(
+			0,
+			// We allow 3 standards per columns
+			standardsConfig.columns * 3,
+		);
+	}
```

Comment:
If I'm honest, this is where I get stuck. From discussing this it is clear that it simply is the case that there is a lot of complex business logic here but I wonder if there is a way to refactor this to be easier to reason about? Of course, asking questions about wondering is very easy!

Are there perhaps too many comments here? In a way I think I am maxed out by trying to read those.

It also doesn't help that it's 17:41 on a Monday and my brain is mush!

Reply:
Yes I can't help but agree - the comments are there to try and explain the business logic but I have a tendancy to overcomment so I will review their usefulness and see if we can get away with less/more concise.

I'll keep staring at this peace of logic and see how it can be tidied up more (if at all) to be more readable.

It is split into essentially 3 sections,
1. 'primaries' (also known as first slice, would this be a better name?)
2. bigs
3. standards

They need to be created in that order as they each have logic which requires the earlier to be known - but perhaps the flow could be more clear!

Reply:
Thanks. But as discussed offline lets focus on moving this one forward, there's a risk you could lose a lot of time refactoring here